gcs/src/main/scala/org/broadinstitute/monster/storage/gcs/GcsApi.scala

```diff
@@ -86,6 +98,261 @@ class GcsApi private[gcs] (runHttp: Request[IO] => Resource[IO, Response[IO]]) {
+          IO.raiseError(
+            new RuntimeException(s"Failed to upload bytes to $path in $bucket")
+          )
+        }
+      }
+    }
+  }
+
+  /**
+    * Initialize a GCS resumable upload.
+    *
+    * Resumable uploads enable pushing data to an object in multiple chunks, and are recommended
+    * for use with files larger than 5MB. The initialization request for the upload must include
+    * any metadata for the object that will be created.
+    *
+    * @see https://cloud.google.com/storage/docs/json_api/v1/how-tos/resumable-upload#start-resumable
+    *
+    * @param bucket the GCS bucket to create the new object within
+    * @param path path within `bucket` where the new object will be created
+    * @param contentType content-type to set on the new object
+    * @param expectedSize total number of bytes that will be uploaded to the new object over
+    *                     the course of all subsequent requests
+    * @param expectedMd5 expected md5, if any, of the new object. Setting this will enable
+    *                    server-side content validation in GCS
+    * @return the unique ID of this upload, for use in subsequent requests
+    */
+  def initResumableUpload(
+    bucket: String,
+    path: String,
+    contentType: `Content-Type`,
+    expectedSize: Long,
+    expectedMd5: Option[String]
+  ): IO[String] = {
+    val objectMetadata = buildUploadMetadata(path, expectedMd5).noSpaces.getBytes
+
+    val gcsReq = Request[IO](
+      method = Method.POST,
+      uri = baseGcsUploadUri(bucket, "resumable"),
+      body = Stream.emits(objectMetadata),
+      headers = Headers.of(
+        `Content-Length`.unsafeFromLong(objectMetadata.length.toLong),
+        `Content-Type`(MediaType.application.json, Charset.`UTF-8`),
+        Header("X-Upload-Content-Length", expectedSize.toString),
+        contentType.toRaw.copy(
+          name = CaseInsensitiveString("X-Upload-Content-Type")
+        )
+      )
+    )
+
+    runHttp(gcsReq).use { response =>
+      response.headers
+        .get(CaseInsensitiveString("X-GUploader-UploadID"))
+        .map { _.value }
+        .liftTo[IO](new RuntimeException(s"no upload token for $path in $bucket"))
+    }
+  }
+
+  /**
+    * Upload a chunk of bytes to an ongoing GCS resumable upload.
+    *
+    * @see https://cloud.google.com/storage/docs/json_api/v1/how-tos/resumable-upload#upload-resumable
+    *
+    * @param bucket the GCS bucket to upload the data into
+    * @param uploadToken the unique ID of the previously-initialized resumable upload
+    * @param rangeStart the (zero-indexed) position of the first byte of `chunkData` within
+    *                   the source file being uploaded
+    * @param chunkData chunk of bytes to upload
+    * @return either the number of bytes stored by GCS so far (signalling that the upload is not complete),
+    *         or Unit when all bytes have been received and GCS signals the upload is done
+    */
+  def uploadChunk(
```

Comment:
Rethinking this name, since the method now supports uploading multiple chunks... Maybe `uploadBytes`?